### DISCUSSION TITLE
Proposal: File and class scoped visibility modifier

### DISCUSSION THREAD

**Discussion by Denis535** [dotnet/csharplang#954] [General]
Kotlin has one good feature: private modifier for classes.

> If you mark a declaration private, it will only be visible inside the file containing the declaration;

https://kotlinlang.org/docs/reference/visibility-modifiers.html

I think it would be very useful for decomposition of classes.
For example: split class into main class and thair helper or into View, Model, Controller.
Now I must to name/rename every class with full name: MyBlaBlaBlaClass, MyBlaBlaBlaClassHelper or MyBlaBlaBlaView, MyBlaBlaBlaModel, MyBlaBlaBlaController. 
With file scoped modifier I could name just: Helper, Model (I'm writting Model and Controller in the same file). 
It will encourage to write more decomposed code and smaller classes.
Only one problem: how those classes can be available for tests.

Also I have one more similar proposal: class scoped modifier (neested classes without real nesting).
For example: MyBlaBlaBlaClass.Helper or MyBlaBlaBlaController.Model.
I don't need to access private members of outer class. I want to avoid hell of long class names and big neested classes.

**Comment by svick** [comment]
> With file scoped modifier I could name just: Helper, Model (I'm writting Model and Controller in the same file).

Why aren't nested classes sufficient for this?

Personally, I like that in C#, files don't really matter, because it makes refactoring easier. This proposal would break that: for example, I can't easily move a file-scoped class to a different file, if it becomes too large.

> Also I have one more similar proposal: class scoped modifier (nested classes without real nesting).

I don't understand what are you proposing here, could could clarify, maybe by showing example code that uses this?

**Comment by Denis535** [comment]
> Why aren't nested classes sufficient for this?

Big neested class or big scope hierarchy makes class terrible and unreadable.

> I don't understand what are you proposing here, could could clarify, maybe by showing example code that uses this?
```
class MyClass {}
class MyClass.Helper {}
```
1) It's very clear that Helper refers to MyClass.
2) When I rename MyClass then MyClass.Helper will be renamed automatically.

**Comment by HaloFour** [comment]
@deni35 

I don't see how that avoids either problem of "long nested class names" or "big nested classes".  You're still embedding the parent class name right in the nested class name.  And as for the size of your nested class, where do you expect that code to go?

If the complaint is that the nested class code is embedded in the same file as the parent class, well, nobody is forcing you to do that.  Define the parent class as a `partial` class and you can have a separate file for the nested class, without trading any of the benefits of it being nested.  At worst you have an extra level of indentation if you go with the default auto-format.

I'm not personally opposed to the notion of class names being compound like that but I think it immediately runs into an ambiguity problem as to whether the nesting is parent classes vs. namespaces.

**Comment by Denis535** [comment]
> long nested class names

Not problem in just long name. Problem in hard name (especially when refactoring, because I can forget to rename other classes) and problem in unreadability neested classes.

> the nested class code is embedded in the same file as the parent class

And it's not problem. I had written in one file:
```
partial class MyController {
    class Model {}
}
partial class MyController {
    private readonly Model model = new Model();
}
```
I just like structured and flat code.

Which option is more readable? It's simply example because of copy past. In real code it's more hard to see where neested example starts or ends.

```
class OuterClass {
    class NestedClass {

        private readonly OuterClass outer;

        public NestedClass(OuterClass outer) {
            this.outer = outer;
        }

        public int Var1 => outer.Var1;
        public int Var2 => outer.Var2;
        public int Var3 => outer.Var3;
        public int Var4 => outer.Var4;
        public int Var5 => outer.Var5;
        public int Var6 => outer.Var6;
        public int Var7 => outer.Var7;
        public int Var8 => outer.Var8;
        public int Var9 => outer.Var9;
        public int Var10 => outer.Var10;


        public int F1() {
            return outer.F1();
        }

        public int F2() {
            return outer.F2();
        }

        public int F3() {
            return outer.F3();
        }
    }

    public int Var1 => 1;
    public int Var2 => 2;
    public int Var3 => 3;
    public int Var4 => 4;
    public int Var5 => 5;
    public int Var6 => 6;
    public int Var7 => 7;
    public int Var8 => 8;
    public int Var9 => 9;
    public int Var10 => 10;


    public int F1() {
        return 1;
    }

    public int F2() {
        return 2;
    }

    public int F3() {
        return 3;
    }
}
```

```
class OuterClass {

    public int Var1 => 1;
    public int Var2 => 2;
    public int Var3 => 3;
    public int Var4 => 4;
    public int Var5 => 5;
    public int Var6 => 6;
    public int Var7 => 7;
    public int Var8 => 8;
    public int Var9 => 9;
    public int Var10 => 10;


    public int F1() {
        return 1;
    }

    public int F2() {
        return 2;
    }

    public int F3() {
        return 3;
    }
}


class OuterClass.NestedClass {

    private readonly OuterClass outer;

    public NestedClass(OuterClass outer) {
        this.outer = outer;
    }

    public int Var1 => outer.Var1;
    public int Var2 => outer.Var2;
    public int Var3 => outer.Var3;
    public int Var4 => outer.Var4;
    public int Var5 => outer.Var5;
    public int Var6 => outer.Var6;
    public int Var7 => outer.Var7;
    public int Var8 => outer.Var8;
    public int Var9 => outer.Var9;
    public int Var10 => outer.Var10;


    public int F1() {
        return outer.F1();
    }

    public int F2() {
        return outer.F2();
    }

    public int F3() {
        return outer.F3();
    }
}
```

**Comment by yaakov-h** [comment]
I'm confused. Your long example above makes no use of `private class` whatsoever.

Swift has `private` and `fileprivate`, and that's been somewhat of a disaster.
